validate-milestone unit gets stuck in infinite loop if LLM fails to call gsd_validate_milestone tool

## Problem
Auto-mode enters an infinite loop when the LLM successfully completes a `validate-milestone` unit conversationally but fails to actually call the `gsd_validate_milestone` tool. 

## Root Cause
In `auto-verification.js` (lines 73-75), `runValidateMilestonePostCheck` returns `"continue"` instead of `"retry"` when the `VALIDATION.md` file does not exist. 

```javascript
// auto-verification.js:70-75
const validationFile = resolveMilestoneFile(s.basePath, mid, "VALIDATION");
if (!validationFile)
    return "continue";
const validationContent = await loadFile(validationFile);
if (!validationContent)
    return "continue"; // <-- Falsely passes the unit
```

Because the unit is marked as passed but no artifact was created, the state machine re-derives the phase as `validating-milestone` and re-dispatches the exact same prompt. Since the LLM receives no feedback that it missed the tool call, it repeats the same conversational output, causing an infinite loop.

## Expected Behavior
When `VALIDATION.md` is missing, the verification check should fail and force a retry, prompting the LLM to actually call the `gsd_validate_milestone` tool.

```javascript
const validationContent = await loadFile(validationFile);
if (!validationContent) {
    s.pendingVerificationRetry = {
        message: "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
        errorClass: "tool-failure"
    };
    return "retry";
}
```

## Environment
- GSD version: 2.82.0
- Model: gpt-5.3-codex
- Unit: validate-milestone/M001

## Reproduction Context
During the completion of milestone `M001`, the model outputted `"Milestone M001 validation complete — verdict: pass."` without calling any tools. The auto-mode loop marked the unit as complete, but since no `VALIDATION.md` was written, the phase remained `validating-milestone`, and it dispatched the same unit 3 times before the stuck-detector caught it.

## Forensic Evidence
- **Anomaly:** `[ERROR] stuck-loop: Unit validate-milestone/M001 was dispatched 3 times`
- **Activity Log `008-validate-milestone-M001.jsonl`:** The assistant generated 15 tokens: `Milestone M001 validation complete — verdict: pass.` without calling `gsd_validate_milestone`.
- **Duration:** The unit "completed" in 1-3 seconds per run.
- **Code:** `runValidateMilestonePostCheck` in `auto-verification.js` returns `"continue"` on missing validation content.

---
*Auto-generated by `/gsd forensics`*

--- a/src/resources/extensions/gsd/auto-verification.ts
+++ b/src/resources/extensions/gsd/auto-verification.ts
@@ -105,22 +105,26 @@ async function runValidateMilestonePostCheck(
   const { milestone: mid } = parseUnitId(s.currentUnit.id);
   if (!mid) return "continue";
 
-  const validationFile = resolveMilestoneFile(s.basePath, mid, "VALIDATION");
-  if (!validationFile) {
+  const setToolFailureRetry = (message: string): VerificationResult => {
     s.pendingVerificationRetry = {
-      message: "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
+      message,
       errorClass: "tool-failure",
     };
     return "retry";
+  };
+
+  const validationFile = resolveMilestoneFile(s.basePath, mid, "VALIDATION");
+  if (!validationFile) {
+    return setToolFailureRetry(
+      "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
+    );
   }
 
   const validationContent = await loadFile(validationFile);
   if (!validationContent) {
-    s.pendingVerificationRetry = {
-      message: "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
-      errorClass: "tool-failure",
-    };
-    return "retry";
+    return setToolFailureRetry(
+      "You must call gsd_validate_milestone to persist the validation results. VALIDATION.md exists but is empty.",
+    );
   }
 
   const verdict = extractVerdict(validationContent);

--- a/src/resources/extensions/gsd/auto-verification.ts
+++ b/src/resources/extensions/gsd/auto-verification.ts
@@ -106,9 +106,13 @@ async function runValidateMilestonePostCheck(
   if (!mid) return "continue";
 
   const setToolFailureRetry = (message: string): VerificationResult => {
+    const retryKey = verificationRetryKey(s.currentUnit!.type, s.currentUnit!.id);
+    const attempt = (s.verificationRetryCount.get(retryKey) ?? 0) + 1;
+    s.verificationRetryCount.set(retryKey, attempt);
     s.pendingVerificationRetry = {
-      message,
-      errorClass: "tool-failure",
+      unitId: s.currentUnit!.id,
+      failureContext: message,
+      attempt,
     };
     return "retry";
   };

--- a/src/resources/extensions/gsd/auto-verification.ts
+++ b/src/resources/extensions/gsd/auto-verification.ts
@@ -106,10 +106,22 @@ async function runValidateMilestonePostCheck(
   if (!mid) return "continue";
 
   const validationFile = resolveMilestoneFile(s.basePath, mid, "VALIDATION");
-  if (!validationFile) return "continue";
+  if (!validationFile) {
+    s.pendingVerificationRetry = {
+      message: "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
+      errorClass: "tool-failure",
+    };
+    return "retry";
+  }
 
   const validationContent = await loadFile(validationFile);
-  if (!validationContent) return "continue";
+  if (!validationContent) {
+    s.pendingVerificationRetry = {
+      message: "You must call gsd_validate_milestone to persist the validation results. No VALIDATION.md was created.",
+      errorClass: "tool-failure",
+    };
+    return "retry";
+  }
 
   const verdict = extractVerdict(validationContent);
   if (verdict !== "needs-remediation") {

--- a/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
+++ b/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
@@ -179,4 +179,26 @@ describe("validate-milestone stuck-loop guard (#4094)", () => {
     assert.match(s.pendingVerificationRetry!.message, /gsd_validate_milestone/);
     assert.equal(s.pendingVerificationRetry!.errorClass, "tool-failure");
   });
+
+  test("retries when VALIDATION file exists but is empty", async () => {
+    insertMilestone({ id: "M001" });
+    insertSlice({ id: "S01", milestoneId: "M001", title: "Slice 1", status: "complete" });
+
+    const path = join(tempDir, ".gsd", "milestones", "M001", "M001-VALIDATION.md");
+    writeFileSync(path, "", "utf-8");
+    invalidateAllCaches();
+
+    const ctx = makeMockCtx();
+    const pi = makeMockPi();
+    const pauseAutoMock = mock.fn(async () => {});
+    const s = makeMockSession(tempDir, "validate-milestone", "M001");
+
+    const result = await runPostUnitVerification({ s, ctx, pi } as VerificationContext, pauseAutoMock);
+
+    assert.equal(result, "retry");
+    assert.equal(pauseAutoMock.mock.callCount(), 0);
+    assert.ok(s.pendingVerificationRetry);
+    assert.match(s.pendingVerificationRetry!.message, /exists but is empty/);
+    assert.equal(s.pendingVerificationRetry!.errorClass, "tool-failure");
+  });
 });

--- a/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
+++ b/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
@@ -176,8 +176,9 @@ describe("validate-milestone stuck-loop guard (#4094)", () => {
     assert.equal(result, "retry");
     assert.equal(pauseAutoMock.mock.callCount(), 0);
     assert.ok(s.pendingVerificationRetry);
-    assert.match(s.pendingVerificationRetry!.message, /gsd_validate_milestone/);
-    assert.equal(s.pendingVerificationRetry!.errorClass, "tool-failure");
+    assert.equal(s.pendingVerificationRetry!.unitId, "M001");
+    assert.match(s.pendingVerificationRetry!.failureContext, /gsd_validate_milestone/);
+    assert.equal(s.pendingVerificationRetry!.attempt, 1);
   });
 
   test("retries when VALIDATION file exists but is empty", async () => {
@@ -198,7 +199,8 @@ describe("validate-milestone stuck-loop guard (#4094)", () => {
     assert.equal(result, "retry");
     assert.equal(pauseAutoMock.mock.callCount(), 0);
     assert.ok(s.pendingVerificationRetry);
-    assert.match(s.pendingVerificationRetry!.message, /exists but is empty/);
-    assert.equal(s.pendingVerificationRetry!.errorClass, "tool-failure");
+    assert.equal(s.pendingVerificationRetry!.unitId, "M001");
+    assert.match(s.pendingVerificationRetry!.failureContext, /exists but is empty/);
+    assert.equal(s.pendingVerificationRetry!.attempt, 1);
   });
 });

--- a/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
+++ b/src/resources/extensions/gsd/tests/validate-milestone-stuck-guard.test.ts
@@ -162,7 +162,7 @@ describe("validate-milestone stuck-loop guard (#4094)", () => {
     assert.equal(pauseAutoMock.mock.callCount(), 0);
   });
 
-  test("continues when no VALIDATION file exists yet", async () => {
+  test("retries when no VALIDATION file exists yet", async () => {
     insertMilestone({ id: "M001" });
     insertSlice({ id: "S01", milestoneId: "M001", title: "Slice 1", status: "complete" });
 
@@ -173,7 +173,10 @@ describe("validate-milestone stuck-loop guard (#4094)", () => {
 
     const result = await runPostUnitVerification({ s, ctx, pi } as VerificationContext, pauseAutoMock);
 
-    assert.equal(result, "continue");
+    assert.equal(result, "retry");
     assert.equal(pauseAutoMock.mock.callCount(), 0);
+    assert.ok(s.pendingVerificationRetry);
+    assert.match(s.pendingVerificationRetry!.message, /gsd_validate_milestone/);
+    assert.equal(s.pendingVerificationRetry!.errorClass, "tool-failure");
   });
 });

